File: abstractions/typescript/src/serialization/serializationWriterFactory.ts
import { SerializationWriter } from './serializationWriter';

/** Defines the contract for a factory that creates SerializationWriter instances. */
export interface SerializationWriterFactory {
    /**
     * Gets the content type this factory creates serialization writers for.
     * @return the content type this factory creates serialization writers for.
     */
    getValidContentType(): string;
    /**
     * Creates a new SerializationWriter instance for the given content type.
     * @param contentType the content type to create a serialization writer for.
     * @return a new SerializationWriter instance for the given content type.
     */
    getSerializationWriter(contentType: string): SerializationWriter;
}

File: abstractions/go/serialization/serialization_writer_factory.go
package serialization

// Defines the contract for a factory that creates SerializationWriter instances.
type SerializationWriterFactory interface {
	// The valid content type for the SerializationWriterFactoryRegistry
	// Returns:
	// 		- the content type
	// 		- nil if the content type is valid
	GetValidContentType() (string, error)
	// Get the relevant SerializationWriter instance for the given content type
	// Parameters:
	// 		- contentType: the content type
	// Returns:
	// 		- the SerializationWriter instance
	//      - an error if any
	GetSerializationWriter(contentType string) (SerializationWriter, error)
}
